src/main/java/com/snowflake/kafka/connector/internal/SnowflakeTelemetryPipeStatus.java

```diff
@@ -0,0 +1,181 @@
+  AtomicLong averageIngestionLagFileCount;              // file count
+  static final String AVERAGE_INGESTION_LAG             = "average_ingestion_lag";
+  static final String AVERAGE_INGESTION_LAG_FILE_COUNT  = "average_ingestion_lag_file_count";
+
+  // Average lag of commit
+  AtomicLong averageCommitLag;                          // average lag between file upload and ingest api calling
+  AtomicLong averageCommitLagFileCount;                 // file count
+  static final String AVERAGE_COMMIT_LAG                = "average_commit_lag";
+  static final String AVERAGE_COMMIT_LAG_FILE_COUNT     = "average_commit_lag_file_count";
+
+  // Need to update two values atomically when calculating lag, thus
+  // a lock is required to protect the access
+  private final Lock lagLock;
+
+  AtomicLong startTime;             // start time of the status recording period
+  static final String START_TIME    = "start_time";
+  static final String END_TIME      = "end_time";
+
+  SnowflakeTelemetryPipeStatus(final String tableName, final String stageName, final String pipeName)
+  {
+    this.tableName = tableName;
+    this.stageName = stageName;
+    this.pipeName = pipeName;
+
+    this.processedOffset = new AtomicLong(0);
+    this.flushedOffset = new AtomicLong(0);
+    this.committedOffset = new AtomicLong(0);
+    this.purgedOffset = new AtomicLong(0);
+    this.totalNumberOfRecord = new AtomicLong(0);
+    this.totalSizeOfData = new AtomicLong(0);
+    this.fileCountOnStage = new AtomicLong(0);
+    this.fileCountOnIngestion = new AtomicLong(0);
+    this.fileCountPurged = new AtomicLong(0);
+    this.fileCountTableStage = new AtomicLong(0);
+    this.cleanerRestartCount = new AtomicLong(0);
+
+    this.averageKafkaLag = new AtomicLong(0);
+    this.averageKafkaLagRecordCount = new AtomicLong(0);
+    this.averageIngestionLag = new AtomicLong(0);
+    this.averageIngestionLagFileCount = new AtomicLong(0);
+    this.averageCommitLag = new AtomicLong(0);
+    this.averageCommitLagFileCount = new AtomicLong(0);
+    this.startTime = new AtomicLong(System.currentTimeMillis());
+
+    this.lagLock = new ReentrantLock();
+  }
+
+  void updateKafkaLag(final long lag)
+  {
+    updateLag(lag, averageKafkaLagRecordCount, averageKafkaLag);
+  }
+
+  void updateIngestionLag(final long lag)
+  {
+    updateLag(lag, averageIngestionLagFileCount, averageIngestionLag);
+  }
+
+  void updateCommitLag(final long lag)
+  {
+    updateLag(lag, averageCommitLagFileCount, averageCommitLag);
+  }
+
+  private void updateLag(final long lag, AtomicLong averageFileCount, AtomicLong averageLag)
+  {
+    lagLock.lock();
+    try {
+      long lagFileCount = averageFileCount.getAndIncrement();
+      // Weighted average to calculate the cumulative average lag
+      averageLag.updateAndGet(value -> (value * lagFileCount + lag) / (lagFileCount + 1));
+    }
+    finally {
+      lagLock.unlock();
+    }
+  }
+
+  boolean empty()
```

Comment:
Is `empty` a common function? if yes, add to `SnowflakeTelemetryBasicInfo`

Reply:
nice catch, will do that